File: clojure/client/tests/raw_websocket_fixture.clj
(ns raw-websocket-fixture
  "Small deterministic RFC 6455 fixture. It validates masked client frames and can
  split UTF-8 text around control frames without hiding behavior in a library."
  (:require [clojure.data.json :as json]
            [clojure.string :as string])
  (:import [java.io ByteArrayOutputStream DataInputStream InputStream OutputStream]
           [java.net InetAddress ServerSocket Socket SocketException]
           [java.nio.charset StandardCharsets]
           [java.security MessageDigest]
           [java.time Duration]
           [java.util Base64]
           [java.util.concurrent Executors TimeUnit]
           [java.util.concurrent.atomic AtomicBoolean AtomicInteger]))

(def ^:private websocket-guid "258EAFA5-E914-47DA-95CA-C5AB0DC85B11")

(declare send-frame!)

(defn- read-http-headers [^InputStream input]
  (let [buffer (ByteArrayOutputStream.)]
    (loop [tail ""]
      (let [byte (.read input)]
        (when (= -1 byte) (throw (java.io.EOFException. "EOF during WebSocket handshake")))
        (.write buffer byte)
        (let [next-tail (str tail (char byte))
              next-tail (subs next-tail (max 0 (- (count next-tail) 4)))]
          (if (= "\r\n\r\n" next-tail)
            (.toString buffer StandardCharsets/US_ASCII)
            (recur next-tail)))))))

(defn- header [headers name]
  (some (fn [line]
          (let [[key value] (string/split line #":" 2)]
            (when (= (string/lower-case key) (string/lower-case name))
              (string/trim value))))
        (string/split-lines headers)))

(defn- accept-value [key]
  (.encodeToString (Base64/getEncoder)
                   (.digest (MessageDigest/getInstance "SHA-1")
                            (.getBytes (str key websocket-guid) StandardCharsets/US_ASCII))))

(defn- handshake! [^Socket socket]
  (let [headers (read-http-headers (.getInputStream socket))
        key (header headers "Sec-WebSocket-Key")]
    (when-not key (throw (AssertionError. "client omitted Sec-WebSocket-Key")))
    (let [response (str "HTTP/1.1 101 Switching Protocols\r\n"
                        "Upgrade: websocket\r\nConnection: Upgrade\r\n"
                        "Sec-WebSocket-Accept: " (accept-value key) "\r\n\r\n")]
      (.write (.getOutputStream socket) (.getBytes response StandardCharsets/US_ASCII))
      (.flush (.getOutputStream socket)))))

(defn- read-byte! [^DataInputStream input]
  (let [value (.read input)]
    (when (= -1 value) (throw (java.io.EOFException. "WebSocket peer closed")))
    value))

(defn read-frame! [connection]
  (let [input ^DataInputStream (:input connection)
        first (read-byte! input)
        second (read-byte! input)
        opcode (bit-and first 0x0f)
        final? (pos? (bit-and first 0x80))
        masked? (pos? (bit-and second 0x80))
        short-length (bit-and second 0x7f)
        length (case short-length
                 126 (.readUnsignedShort input)
                 127 (.readLong input)
                 short-length)]
    (when-not masked? (throw (AssertionError. "client WebSocket frame was not masked")))
    (when (> length (* 2 1024 1024)) (throw (AssertionError. "fixture frame too large")))
    (let [mask (byte-array 4)
          payload (byte-array (int length))]
      (.readFully input mask)
      (.readFully input payload)
      (dotimes [index (alength payload)]
        (aset-byte payload index
                   (byte (bit-xor (bit-and 0xff (aget payload index))
                                  (bit-and 0xff (aget mask (mod index 4)))))))
      {:opcode opcode :final? final? :payload payload})))

(defn read-message! [connection]
  (let [parts (ByteArrayOutputStream.)]
    (loop [first? true]
      (let [{:keys [opcode final? payload]} (read-frame! connection)]
        (cond
          (= opcode 8) (throw (java.io.EOFException. "client closed WebSocket"))
          (= opcode 9) (do (send-frame! connection 10 true payload) (recur first?))
          (= opcode 10) (recur first?)
          (and first? (not= opcode 1)) (throw (AssertionError. (str "expected text frame, got opcode " opcode)))
          (and (not first?) (not= opcode 0)) (throw (AssertionError. "expected continuation frame"))
          :else (do (.write parts payload)
                    (if final?
                      (json/read-str (.toString parts StandardCharsets/UTF_8))
                      (recur false))))))))

(defn send-frame! [connection opcode final? payload]
  (let [bytes (if (string? payload) (.getBytes ^String payload StandardCharsets/UTF_8) payload)
        output ^OutputStream (:output connection)
        first (bit-or opcode (if final? 0x80 0))]
    (.write output first)
    (cond
      (< (alength ^bytes bytes) 126) (.write output (alength ^bytes bytes))
      (<= (alength ^bytes bytes) 65535) (do (.write output 126)
                                            (.write output (bit-and 0xff (bit-shift-right (alength ^bytes bytes) 8)))
                                            (.write output (bit-and 0xff (alength ^bytes bytes))))
      :else (do
              (.write output 127)
              (doseq [shift (range 56 -1 -8)]
                (.write output (bit-and 0xff (bit-shift-right (long (alength ^bytes bytes)) shift))))))
    (.write output ^bytes bytes)
    (.flush output)))

(defn send-text! [connection value]
  (send-frame! connection 1 true (if (string? value) value (json/write-str value))))

(defn send-fragmented-text! [connection value split-at]
  (let [bytes (.getBytes ^String (if (string? value) value (json/write-str value)) StandardCharsets/UTF_8)
        first (java.util.Arrays/copyOfRange bytes 0 split-at)
        second (java.util.Arrays/copyOfRange bytes split-at (alength bytes))]
    (send-frame! connection 1 false first)
    ;; A control frame between fragments proves parser state survives it.
    (send-frame! connection 9 true (.getBytes "ping" StandardCharsets/UTF_8))
    (send-frame! connection 0 true second)))

(defn close-transport! [connection] (.close ^Socket (:socket connection)))

(defrecord Fixture [^ServerSocket server executor stopped connections failure handler]
  java.lang.AutoCloseable
  (close [_]
    (.set ^AtomicBoolean stopped true)
    (.close server)
    (.shutdownNow executor)
    (.awaitTermination executor 2 TimeUnit/SECONDS)
    (when-let [error @failure] (throw error))))

(defn fixture [handler]
  (let [server (ServerSocket. 0 20 (InetAddress/getLoopbackAddress))
        executor (Executors/newCachedThreadPool)
        stopped (AtomicBoolean. false)
        connections (AtomicInteger. 0)
        failure (atom nil)
        result (->Fixture server executor stopped connections failure handler)]
    (.execute executor
              (reify Runnable
                (run [_]
                  (while (not (.get stopped))
                    (try
                      (let [socket (.accept server)
                            index (.getAndIncrement connections)]
                        (.setSoTimeout socket 10000)
                        (.execute executor
                                  (reify Runnable
                                    (run [_]
                                      (with-open [owned socket]
                                        (try
                                          (handshake! socket)
                                          (handler {:socket socket
                                                    :input (DataInputStream. (.getInputStream socket))
                                                    :output (.getOutputStream socket)} index)
                                          (catch SocketException _ nil)
                                          (catch java.io.EOFException _ nil)
                                          (catch InterruptedException _
                                            (.interrupt (Thread/currentThread)))
                                          (catch Throwable error
                                            (compare-and-set! failure nil error))))))))
                      (catch SocketException _ nil))))))
    result))

(defn url [fixture] (str "http://127.0.0.1:" (.getLocalPort ^ServerSocket (:server fixture))))

(defn version [query-set timestamp]
  {"querySet" query-set "identity" 0
   "ts" (if (zero? timestamp) "AAAAAAAAAAA=" (str "fixture-ts-" timestamp))})

(defn updated [query-id value & [logs]]
  (cond-> {"type" "QueryUpdated" "queryId" query-id "value" value}
    (seq logs) (assoc "logLines" logs)))

(defn failed [query-id]
  {"type" "QueryFailed" "queryId" query-id "errorMessage" "fixture failure"
   "errorData" {"code" "FIXTURE"}})

(defn transition [start end & modifications]
  {"type" "Transition" "startVersion" start "endVersion" end
   "modifications" (vec modifications)})

File: kotlin/client/tests/fixtures/RawWebSocketFixture.kt
package convex.kotlin.testing

import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.EOFException
import java.io.IOException
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketTimeoutException
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.time.Duration
import java.util.Base64
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * A deliberately tiny RFC 6455 peer for exercising the JDK WebSocket client.
 * It speaks raw frames so tests can split UTF-8, stall mid-frame, and terminate
 * transport without another WebSocket implementation hiding those details.
 */
class RawWebSocketFixture(
    private val handler: (Connection, Int) -> Unit,
) : AutoCloseable {
    private val server = ServerSocket(0, 50, java.net.InetAddress.getLoopbackAddress())
    private val workers =
        Executors.newCachedThreadPool { task ->
            Thread(task, "raw-websocket-fixture").apply { isDaemon = true }
        }
    private val failures = ConcurrentLinkedQueue<Throwable>()
    private val connectionCounter = AtomicInteger()
    private val accepted = CountDownLatch(1)

    val url = "http://127.0.0.1:${server.localPort}"

    init {
        workers.execute {
            while (!server.isClosed) {
                try {
                    val socket = server.accept()
                    val index = connectionCounter.getAndIncrement()
                    accepted.countDown()
                    workers.execute {
                        try {
                            Connection(socket).use { connection ->
                                connection.handshake()
                                handler(connection, index)
                            }
                        } catch (error: Throwable) {
                            if (!server.isClosed && error !is IOException && error !is InterruptedException) {
                                failures += error
                            }
                        }
                    }
                } catch (error: Throwable) {
                    if (!server.isClosed) failures += error
                }
            }
        }
    }

    fun awaitConnection(timeout: Duration = Duration.ofSeconds(3)) =
        check(accepted.await(timeout.toMillis(), TimeUnit.MILLISECONDS)) {
            "WebSocket client did not connect"
        }

    fun assertHealthy() {
        failures.poll()?.let { throw AssertionError("raw WebSocket fixture failed", it) }
    }

    override fun close() {
        server.close()
        workers.shutdownNow()
        workers.awaitTermination(2, TimeUnit.SECONDS)
        assertHealthy()
    }

    class Connection internal constructor(
        private val socket: Socket,
    ) : AutoCloseable {
        private val input = DataInputStream(socket.getInputStream())
        private val output = socket.getOutputStream()

        internal fun handshake() {
            val request = readHttpHeaders()
            val key =
                request
                    .lineSequence()
                    .firstOrNull { it.startsWith("Sec-WebSocket-Key:", ignoreCase = true) }
                    ?.substringAfter(':')
                    ?.trim()
                    ?: error("handshake omitted Sec-WebSocket-Key")
            val accept =
                Base64.getEncoder().encodeToString(
                    MessageDigest
                        .getInstance("SHA-1")
                        .digest("$key$WEBSOCKET_GUID".toByteArray(StandardCharsets.US_ASCII)),
                )
            output.write(
                (
                    "HTTP/1.1 101 Switching Protocols\r\n" +
                        "Upgrade: websocket\r\n" +
                        "Connection: Upgrade\r\n" +
                        "Sec-WebSocket-Accept: $accept\r\n\r\n"
                ).toByteArray(StandardCharsets.US_ASCII),
            )
            output.flush()
        }

        fun readText(timeout: Duration = Duration.ofSeconds(3)): String {
            socket.soTimeout = timeout.toMillis().toInt()
            val message = ByteArrayOutputStream()
            var started = false
            while (true) {
                val frame = readFrame()
                when (frame.opcode) {
                    0x1 -> {
                        check(!started) { "nested text frame" }
                        started = true
                        message.write(frame.payload)
                    }
                    0x0 -> {
                        check(started) { "continuation without text frame" }
                        message.write(frame.payload)
                    }
                    0x8 -> throw EOFException("client closed WebSocket")
                    0x9 -> sendFrame(0xA, frame.payload)
                    0xA -> continue
                    else -> error("unexpected client opcode ${frame.opcode}")
                }
                if (started && frame.final) {
                    return message.toByteArray().toString(StandardCharsets.UTF_8)
                }
            }
        }

        @Synchronized
        fun sendText(text: String) = sendFrame(0x1, text.toByteArray(StandardCharsets.UTF_8))

        /** Split inside a multibyte code point and interleave a control frame. */
        @Synchronized
        fun sendFragmentedText(
            text: String,
            firstFragmentBytes: Int,
        ) {
            val bytes = text.toByteArray(StandardCharsets.UTF_8)
            require(firstFragmentBytes in 1 until bytes.size)
            sendFrame(0x1, bytes.copyOfRange(0, firstFragmentBytes), final = false, flush = false)
            sendFrame(0x9, "probe".toByteArray(), flush = false)
            sendFrame(0x0, bytes.copyOfRange(firstFragmentBytes, bytes.size), flush = true)
        }

        @Synchronized
        fun sendPing(payload: String = "ping") = sendFrame(0x9, payload.toByteArray())

        /** Consume a frame prefix, then leave the peer blocked in its payload parser. */
        @Synchronized
        fun sendPartialText(
            declaredLength: Int,
            prefix: ByteArray,
        ) {
            require(declaredLength in 1..125 && prefix.size < declaredLength)
            output.write(byteArrayOf(0x81.toByte(), declaredLength.toByte()))
            output.write(prefix)
            output.flush()
        }

        fun closeTransport() = socket.close()

        /** Force an I/O reset instead of performing a cooperative WebSocket close. */
        fun abortTransport() {
            socket.setSoLinger(true, 0)
            socket.close()
        }

        fun awaitClientClose(timeout: Duration = Duration.ofSeconds(3)) {
            socket.soTimeout = timeout.toMillis().toInt()
            try {
                while (input.read() != -1) {
                    // An abort can arrive after queued bytes. Drain until EOF.
                }
            } catch (_: SocketTimeoutException) {
                error("client did not retire the old connection")
            }
        }

        override fun close() = socket.close()

        private fun readHttpHeaders(): String {
            val bytes = ByteArrayOutputStream()
            var matched = 0
            val terminator = byteArrayOf(13, 10, 13, 10)
            while (matched < terminator.size) {
                val byte = input.read()
                if (byte < 0) throw EOFException("handshake ended early")
                bytes.write(byte)
                matched = if (byte.toByte() == terminator[matched]) matched + 1 else 0
                check(bytes.size() <= 32 * 1024) { "oversized HTTP handshake" }
            }
            return bytes.toString(StandardCharsets.US_ASCII)
        }

        private fun readFrame(): Frame {
            val first = input.readUnsignedByte()
            val second = input.readUnsignedByte()
            val final = first and 0x80 != 0
            val opcode = first and 0x0f
            val masked = second and 0x80 != 0
            var length = second and 0x7f
            if (length == 126) length = input.readUnsignedShort()
            if (length == 127) {
                val longLength = input.readLong()
                require(longLength <= Int.MAX_VALUE) { "fixture frame is too large" }
                length = longLength.toInt()
            }
            check(masked) { "client frames must be masked" }
            val mask = ByteArray(4).also(input::readFully)
            val payload = ByteArray(length).also(input::readFully)
            payload.indices.forEach { index ->
                payload[index] = (payload[index].toInt() xor mask[index % 4].toInt()).toByte()
            }
            return Frame(final, opcode, payload)
        }

        private fun sendFrame(
            opcode: Int,
            payload: ByteArray,
            final: Boolean = true,
            flush: Boolean = true,
        ) {
            output.write((if (final) 0x80 else 0x00) or opcode)
            when {
                payload.size < 126 -> output.write(payload.size)
                payload.size <= 65_535 -> {
                    output.write(126)
                    output.write(payload.size ushr 8)
                    output.write(payload.size and 0xff)
                }
                else -> error("fixture payload too large")
            }
            output.write(payload)
            if (flush) output.flush()
        }
    }

    private data class Frame(
        val final: Boolean,
        val opcode: Int,
        val payload: ByteArray,
    )

    private companion object {
        const val WEBSOCKET_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
    }
}
